### PR TITLE
Fix readexactly eof

### PR DESCRIPTION
**Pull request by SafaSafari**
## What do these changes do?
fix readexactly last chunk with eof and make the user decide to select between readexactly and read

## Are there changes in behavior for the user?
maybe yes

## Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by webknjaz** [comment]
Hi, such a change would require a test and a more detailed explanation of the effect of applying it. Saying that it fixes something and not explaining what was broken and how the behavior is affected, and what's the difference, makes it hard for us to judge how dangerous this patch is and if it's even necessary.
Could you please fix this?

**Comment by SafaSafari** [comment]
> Hi, such a change would require a test and a more detailed explanation of the effect of applying it. Saying that it fixes something and not explaining what was broken and how the behavior is affected, and what's the difference, makes it hard for us to judge how dangerous this patch is and if it's even necessary.
> Could you please fix this?

yeah
let me explain the problem in this video

output

**Comment by SafaSafari** [comment]
> Hi, such a change would require a test and a more detailed explanation of the effect of applying it. Saying that it fixes something and not explaining what was broken and how the behavior is affected, and what's the difference, makes it hard for us to judge how dangerous this patch is and if it's even necessary. Could you please fix this?

yeah, you're right
let me explain in a video

output

**Comment by SafaSafari** [comment]
it's for the main problem about readexactly function
I also modified iter_chunked function to make the user decide between read and readexaclty
because when we use iter_chunked function, we expect every chunk size exactly equal to input int variable

**Comment by Dreamsorcerer** [comment]
> yeah, you're right let me explain in a video

You just show the change, not the reason. Most often, a chunk size is to set an upper limit on how much data is read into memory, so the application doesn't use too much. Why do we want to stop the application processing data if it is available? Your change merely delays any available data from reaching the application. Why do we want to return something at EOF instead of producing an error? Your change results in `readexactly()` returning data which is _not_ exactly the requested length.

In other words, what are the use cases that justify this change?

**Comment by SafaSafari** [comment]
So let me explain the problem from the beginning
I need to divide a large file into equal chunks using iter_chunked
But it depends on network speed and yielding chunks lower than chunk_size
I opened 2 pull requests before and changed the read function to return exactly the input size
A person said we have readexactly and I've seen that rating an exception
So do this commit
I think I should create another function without editing readexactly 🤔

**Comment by Dreamsorcerer** [comment]
> I need to divide a large file into equal chunks using iter_chunked

I'm not sure why you would need to use aiohttp's iter_chunked to do this, it's still unclear what the overall goal is. It's sounding like a rather obscure use case. We probably don't want to merge something into this library unless it's going to be useful for a significant number of users (or atleast is very difficult to do without changes to the library).

For a use case like this, it is very easy to handle in the application code. e.g. Something like:
```
result = b""
async for chunk in iter_chunked():
    result += chunk
    while len(result) > part_size:
        save_part(result[:part_size])
        result = result[part_size:]
if result:
    save_part(result)
```
So, we're just not really seeing a reason this should be added to the library for everyone, if it might only be used by you.

**Comment by SafaSafari** [comment]
Sure
